### PR TITLE
Unit tests for component set utils

### PR DESCRIPTION
**Pull request by jeffb-sfdc**
### What does this PR do?
* Adds unit tests for PR https://github.com/forcedotcom/salesforcedx-vscode/pull/4891
* Adds a debug launch condition for VSCode Core
* Adds Jest dependencies to core's package.json

### What issues does this PR fix or reference?
https://github.com/forcedotcom/salesforcedx-vscode/issues/3663, @W-10273533@

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -210,7 +209,7 @@
         "setup-and-run",
         "-s",
         "~/.vscode-test-resources",
-        "-o", 
+        "-o",
```
Comment (jeffb-sfdc) [inline .vscode/launch.json:212]: removed extra whitespace (was done automatically)

File: .vscode/launch.json [hunk 2]
```diff
@@ -622,22 +621,31 @@
...
-        "cwd": "${workspaceRoot}/packages/salesforcedx-sobjects-faux-generator",
-        "type": "node",
-        "request": "launch",
-        "runtimeArgs": ["--inspect-brk", "${workspaceRoot}/packages/salesforcedx-sobjects-faux-generator/node_modules/.bin/jest", "--runInBand", "--coverage", "false"],
-        "console": "integratedTerminal",
-        "internalConsoleOptions": "neverOpen"
-    }, 
-    {
-      "name": "Debug Apex Debug Jest Unit Tests",
+      "name": "Debug Tests - Apex Debug - Unit Tests",
```
Comment (jeffb-sfdc) [inline .vscode/launch.json:624]: Standardized the naming of the unit test configs to mach the "Launch Tests" configs.

File: .vscode/launch.json [hunk 3]
```diff
@@ -622,22 +621,31 @@
...
       "cwd": "${workspaceRoot}/packages/salesforcedx-apex-debugger",
       "type": "node",
       "request": "launch",
       "runtimeArgs": ["--inspect-brk", "${workspaceRoot}/packages/salesforcedx-apex-debugger/node_modules/.bin/jest", "--runInBand", "--coverage", "false"],
       "console": "integratedTerminal",
       "internalConsoleOptions": "neverOpen"
+    },
+    {
+      "name": "Debug Tests - VSCode Core - Unit Tests",
```
Comment (jeffb-sfdc) [inline .vscode/launch.json:633]: Added new "Debug Tests - VSCode Core - Unit Tests" launch config.

File: packages/salesforcedx-vscode-apex-debugger/test/jest/context/isvContext.test.ts [hunk 4]
```diff
@@ -18,7 +18,7 @@ describe('isvContext unit test', () => {
       onDidChangeSpy = jest.fn();
       onDidCreateSpy = jest.fn();
       onDidDeleteSpy = jest.fn();
-      // explictly set this return value due to the VS Code mock not being reset
+      // explicitly set this return value due to the VS Code mock not being reset
```
Comment (jeffb-sfdc) [inline packages/salesforcedx-vscode-apex-debugger/test/jest/context/isvContext.test.ts:21]: typo

File: packages/salesforcedx-vscode-core/package.json [hunk 5]
```diff
@@ -47,6 +47,7 @@
     "@types/adm-zip": "^0.5.0",
     "@types/chai": "4.3.3",
     "@types/glob": "^7.2.0",
+    "@types/jest": "28.1.7",
```
Comment (jeffb-sfdc) [inline packages/salesforcedx-vscode-core/package.json:50]: what was weird, is that jest was in vscode-core's `scripts`, and one could run jest against vscode-core, but vscode-core's dependencies didn't include Jest.  I added them to match other projects.

File: packages/salesforcedx-vscode-core/src/services/sdr/componentSetUtils.ts [hunk 6]
```diff
@@ -24,9 +24,11 @@ export async function setApiVersion(componentSet: ComponentSet): Promise<void> {
     return;
   }
 
-  // If no user-configured Api Version is present, then get the version from the Org.
-  const orgApiVersion = await getOrgApiVersion();
-  componentSet.apiVersion = orgApiVersion ?? componentSet.apiVersion;
+  // If no user-configured API Version is present, then get the version from the org.
+  const orgApiVersion = await exports.getOrgApiVersion();
+  componentSet.apiVersion = orgApiVersion && orgApiVersion.length > 0
```
Comment (jeffb-sfdc) [inline packages/salesforcedx-vscode-core/src/services/sdr/componentSetUtils.ts:29]: It turns out that `orgApiVersion ?? componentSet.apiVersion` doesn't work for an empty string.  (found this out while writing the unit tests, so unit tests "For The Win", eh?

File: packages/salesforcedx-vscode-core/src/services/sdr/componentSetUtils.ts [hunk 7]
```diff
@@ -52,15 +54,16 @@ export async function setSourceApiVersion(componentSet: ComponentSet): Promise<v
 
   // Next, if it still is not set, set it to the highest API version supported by the target org.
   if (!sourceApiVersion) {
-    const orgApiVersion = await getOrgApiVersion();
-    sourceApiVersion = orgApiVersion ?? componentSet.sourceApiVersion;
+    const orgApiVersion = await exports.getOrgApiVersion();
```
Comment (jeffb-sfdc) [inline packages/salesforcedx-vscode-core/src/services/sdr/componentSetUtils.ts:57]: When one has:
someFile.ts
```
export function a() {
  b();
}

export function b() {
}
```

When writing unit tests for someFile.ts, one isn't able to mock `b()`.  In order to mock `b()`, one needs to add `'exports.'`, like so:
```
export function a() {
  exports.b();
}

export function b() {
}
```

File: packages/salesforcedx-vscode-core/test/jest/services/sdr/componentSetUtils.test.ts [hunk 8]
```diff
@@ -4,61 +4,187 @@
  * Licensed under the BSD 3-Clause license.
  * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
+
```
Comment (jeffb-sfdc) [inline packages/salesforcedx-vscode-core/test/jest/services/sdr/componentSetUtils.test.ts:7]: componentSetUtils now has 100% coverage

Comment (jeffb-sfdc) [reply packages/salesforcedx-vscode-core/test/jest/services/sdr/componentSetUtils.test.ts:7]: Screenshot 2023-06-09 at 9 08 28 AM

File: packages/salesforcedx-vscode-core/src/services/sdr/componentSetUtils.ts [hunk 9]
```diff
@@ -52,15 +54,16 @@ export async function setSourceApiVersion(componentSet: ComponentSet): Promise<v
...
+    sourceApiVersion = orgApiVersion;
```
Comment (RitamAgrawal) [inline packages/salesforcedx-vscode-core/src/services/sdr/componentSetUtils.ts:58]: Is there a possibility that componentSet already has sourceApiVersion set when passed through this function?

File: packages/salesforcedx-vscode-core/test/jest/services/sdr/componentSetUtils.test.ts [hunk 10]
```diff
@@ -4,61 +4,187 @@
...
+      await componentSetUtils.setSourceApiVersion(componentSet);
+
+      expect(componentSet.sourceApiVersion).toBe('53.0');
+      expect(getValueMock).toHaveBeenCalled();
+      expect(getUserConfiguredApiVersionMock).not.toHaveBeenCalled();
+      expect(getOrgApiVersionMock).not.toHaveBeenCalled();
+    });
+
+    it('should validate that sourceApiVersion is set from getUserConfiguredApiVersion', async () => {
```
Comment (RitamAgrawal) [inline packages/salesforcedx-vscode-core/test/jest/services/sdr/componentSetUtils.test.ts:124]: May be just add 'when not set (undefined) through componentSet or SfdxProjectConfig'

**Review by RitamAgrawal — COMMENTED** [review]
LGTM!

**Comment by RitamAgrawal** [comment]
QA Notes:
✅ New Debug config added
✅ Code coverage - 100%
✅ Tests cover the order of precedence to be followed for the value of sourceApiVersion

File: packages/salesforcedx-vscode-core/src/services/sdr/componentSetUtils.ts (hunk 9 above)
Comment (jeffb-sfdc) [reply packages/salesforcedx-vscode-core/src/services/sdr/componentSetUtils.ts:58]: I'm not sure I understand the question.  When `setSourceApiVersion(componentSet: ComponentSet)` is called...
...in the scenario that the sourceApiVersion has been set via a metadata file, by the time `setSourceApiVersion()` is called, `componentSet.sourceApiVersion` has been set, and the function exits at line 43.

...in the scenario that the sourceApiVersion is not set, it wasn't set via a metadata file, and then we try to get it from sfdx-project.json, etc...

Let's discuss in our standup/huddle this morning.

File: packages/salesforcedx-vscode-core/test/jest/services/sdr/componentSetUtils.test.ts (hunk 10 above)
Comment (jeffb-sfdc) [reply packages/salesforcedx-vscode-core/test/jest/services/sdr/componentSetUtils.test.ts:124]: done

**Review by RitamAgrawal — APPROVED** [review]

**Pull request merged**